### PR TITLE
[NFC][X86] Reorder the registers to reduce unnecessary iterations

### PR DESCRIPTION
**Pull request by KanRobert**
* Introduce field `PositionOrder` for class `Register` and `RegisterTuples`
* If register A's `PositionOrder` < register B's `PositionOrder`, then A is placed before B in the enum in X86GenRegisterInfo.inc
* The new order of registers in the enum for X86 will be
      1. Registers before AVX512,
      2. AVX512 registers (X/YMM16-31, ZMM0-31, K registers)
      3. AMX registers (TMM)
      4.  APX registers (R16-R31)
* Add a new target hook `getNumSupportedRegs()` to return the number of registers for the function (may overestimate). 
* Replace `getNumRegs()` with  `getNumSupportedRegs()` in LiveVariables to eliminate iterations on unsupported registers

This patch can reduce 0.3% instruction count regression for sqlite3 during compile-stage (O3) by not iterating on APX registers
for #67702

### CODE REVIEW

**Comment by RKSimon** [comment]
Please can you fix the clang-format fails

File: llvm/utils/TableGen/CodeGenRegisters.cpp [hunk 1]
```diff
@@ -1175,24 +1175,21 @@ CodeGenRegBank::CodeGenRegBank(RecordKeeper &Records,
...
-    for (Record *RC : TupRegs)
-      getReg(RC);
+    Regs.insert(Regs.end(), TupRegs.begin(), TupRegs.end());
   }
 
+  llvm::sort(Regs, LessRecordRegister());
+  // Assign the enumeration values.
+  for (unsigned i = 0, e = Regs.size(); i != e; ++i)
+    getReg(Regs[i]);
+
```
Comment (phoebewang) [inline llvm/utils/TableGen/CodeGenRegisters.cpp:1192]: Why switch the order? This doesn't seem related to your change.

**Comment by KanRobert** [comment]
> Please can you fix the clang-format fails

Done.

File: llvm/utils/TableGen/CodeGenRegisters.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/utils/TableGen/CodeGenRegisters.cpp]: It's related. `Register` and `RegisterTuple` are all added to same enum in X86GenRegisterInfo.inc. So we need to merge the vectors and sort only once.

File: llvm/lib/Target/X86/X86RegisterInfo.cpp [hunk 2]
```diff
@@ -616,6 +616,32 @@ BitVector X86RegisterInfo::getReservedRegs(const MachineFunction &MF) const {
...
+  bool HasAVX = ST.hasAVX();
+  bool HasAVX512 = ST.hasAVX512();
+  bool HasAMX = ST.hasAMXTILE();
+  if (HasAMX)
+    return X86::TMM7 + 1;
+  if (HasAVX512)
+    return X86::K6_K7 + 1;
+  if (HasAVX)
+    return X86::YMM15 + 1;
+  return X86::YMM0;
```
Comment (phoebewang) [inline llvm/lib/Target/X86/X86RegisterInfo.cpp:643]: Is there any way to avoid to hard code it?
If not, I think you'd better to add assert to make sure the number is the boundary, e.g.,

`assert(X86::YMM15 + 1 == X86:K0, "Register number may not correct");`

File: llvm/lib/Target/X86/X86RegisterInfo.cpp [hunk 3]
```diff
@@ -616,6 +616,32 @@ BitVector X86RegisterInfo::getReservedRegs(const MachineFunction &MF) const {
...
+unsigned X86RegisterInfo::getNumSupportedRegs(const MachineFunction &MF) const {
+  const X86Subtarget &ST = MF.getSubtarget<X86Subtarget>();
+  // All existing Intel CPUs that support AMX support AVX512 and all existing
+  // Intel CPUs that support APX support AMX. AVX512 implies AVX.
+  //
+  // We enumerate the registers in X86GenRegisterInfo.inc in this order:
+  //
+  // Registers before AVX512,
+  // AVX512 registers (X/YMM16-31, ZMM0-31, K registers)
+  // AMX registers (TMM)
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86RegisterInfo.cpp:629]: We can also see the difference by the update in llvm/test/CodeGen/X86/ipra-reg-usage.ll

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h [hunk 4]
```diff
@@ -1182,11 +1182,27 @@ namespace X86II {
     }
   }
 
+  inline bool isXMMReg(unsigned RegNo) {
```
Comment (phoebewang) [inline llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:1185]: Add doxygen for them.

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h [hunk 5]
```diff
@@ -1182,11 +1182,27 @@ namespace X86II {
...
+    return (RegNo >= X86::XMM0 && RegNo <= X86::XMM15) ||
```
Comment (phoebewang) [inline llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:1186]: Add comment to explain the reason, besides, add assert to make sure they are continuous, e.g.,

`assert(X86::XMM15 - X86::XMM0 == 15, "XMM registers are not continuous");`

Comment (phoebewang) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:1186]: Besides, this will bring problems to code like this: https://github.com/llvm/llvm-project/blob/main/llvm/lib/Target/X86/X86RegisterInfo.cpp#L608

File: llvm/lib/Target/X86/X86RegisterInfo.cpp (hunk 2 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86RegisterInfo.cpp:648]: Done

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h (hunk 4 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:1186]: Done

File: llvm/lib/CodeGen/LiveVariables.cpp [hunk 6]
```diff
@@ -406,11 +406,11 @@ bool LiveVariables::HandlePhysRegKill(Register Reg, MachineInstr *MI) {
   return true;
 }
 
-void LiveVariables::HandleRegMask(const MachineOperand &MO) {
+void LiveVariables::HandleRegMask(const MachineOperand &MO, unsigned NumRegs) {
   // Call HandlePhysRegKill() for all live registers clobbered by Mask.
   // Clobbered registers are always dead, sp there is no need to use
   // HandlePhysRegDef().
-  for (unsigned Reg = 1, NumRegs = TRI->getNumRegs(); Reg != NumRegs; ++Reg) {
+  for (unsigned Reg = 1; Reg != NumRegs; ++Reg) {
```
Comment (topperc) [inline llvm/lib/CodeGen/LiveVariables.cpp:413]: I want to take a step back.

Have you tried checking MRI->isReserved(Reg) here? Does that do anything to improve compile time?

Where is the time spent?

Am I reading this right, that it calls HandlePhysRegKill on the super register anytime it finds one of the subregisters is clobbered? Even if its already called it before?

File: llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h (hunk 5 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/MCTargetDesc/X86BaseInfo.h:1191]: We also updated the code there.

File: llvm/lib/CodeGen/LiveVariables.cpp (hunk 6 above)
Comment (KanRobert) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: I did an analysis on the spent time there https://github.com/llvm/llvm-project/pull/67702#issuecomment-1767656271.

The extra time is spent on iterations over the new registers and new register classes. LiveVariables is one of the passes with the greatest impact b/c it iterates them for each basic block. Besides, for LiveVariables and some other passes, the new registers make a bigger regmask and regset. The operations on the regmask and regset take more time too, including allocation and loopup, etc.

`MRI->isReserved(Reg)`? I think it's a AArch64 function only and even not used by X86 at all.

> Am I reading this right, that it calls HandlePhysRegKill on the super register anytime it finds one of the subregisters is clobbered? Even if its already called it before?

My understanding is same as yours.

Comment (topperc) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: MRI->isReserved(Reg) doesn't use the information from X86RegisterInfo::getReservedRegs?

Comment (KanRobert) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: I misunderstood your idea...

diff --git a/llvm/lib/CodeGen/LiveVariables.cpp b/llvm/lib/CodeGen/LiveVariables.cpp
index 6b983b6320c7..44e127151004 100644
--- a/llvm/lib/CodeGen/LiveVariables.cpp
+++ b/llvm/lib/CodeGen/LiveVariables.cpp
@@ -411,6 +411,9 @@ void LiveVariables::HandleRegMask(const MachineOperand &MO) {
   // Clobbered registers are always dead, sp there is no need to use
   // HandlePhysRegDef().
   for (unsigned Reg = 1, NumRegs = TRI->getNumRegs(); Reg != NumRegs; ++Reg) {
+    // Skip reserved regs.
+    if (MRI->isReserved(Reg))
+      continue;
     // Skip dead regs.
     if (!PhysRegDef[Reg] && !PhysRegUse[Reg])
       continue;

After testing, I find checking `MRI->isReserved(Reg)` does not improve compile time, instead, it will bring 0.2% instruction count regression.

Comment (KanRobert) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: What do you mean by "deprecated"? I see this function is still called when compiling workloads.

Comment (arsenm) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: As in all uses of it should be removed and replaced with LiveIntervals. You still see it because nobody ever completed the work to remove it

Comment (KanRobert) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: I see. But currently LiveVariables brings about 0.3% instruction count regression when I introduce new registers in #67702. If the regression is intolerable, then it will block my APX enabling work. I have to resolve it to before the migration is complete.

**Comment by KanRobert** [comment]
Ping?

File: llvm/lib/CodeGen/LiveVariables.cpp (hunk 6 above)
Comment (topperc) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: How much work is left to remove LIveVariables?

Comment (KanRobert) [reply llvm/lib/CodeGen/LiveVariables.cpp:413]: @jayfoad commented in #67702. I believe he/she knows more about that than me.

**Review by phoebewang — APPROVED** [review]
I don't see objections to this patch, so I'll approve to unblock our follow up patches.
I think as a workaround, this patch solves the compiler time regression caused by adding new registers. We can remove it once we don't need LIveVariables.
